Implement BitmapData.perlinNoise()

torokati44 opened the pull request:
This aims to help with #2240, although that is still blocked by `BitmapData.draw` (also by `merge`, but that doesn't seem to be enough in itself).
Also should improve #2295, but without `BitmapData.paletteMap`, it actually regresses a bit - at least visually.

I tested these changes with lots of different parameter combinations, using [this ActionScript code](https://gist.github.com/torokati44/c1b5a859d7c96bb4022ac36ed493b2f4), which I compiled with `mtasc` into [this .swf](https://github.com/ruffle-rs/ruffle/files/5865428/BitmapPerlinNoise.zip).

This produces the following image:

![image](https://user-images.githubusercontent.com/288816/105685618-903fd400-5ef6-11eb-9bec-dd9023cd5be4.png)

**EDIT:**
I managed to "fix" the rounding when converting the floating point values to `u8` in all cases.
Now, when taking the example `.swf` above, and making four variations of it: Setting the `transparency` property of the `BitmapData`; and the `fractalNoise` parameter of `perlinNoise()` to be always `true`/`false` (independent of each other); every single pixel matches perfectly between Ruffle and Flash Player in all four cases.

As usual, there are some ways performance could be improved, but for now, this is more than nothing.

Herschel commented on core/src/turbulence.rs:
```diff
@@ -0,0 +1,240 @@
+/// This file is a Rust port of the C reference implementation of the
+/// feTurbulence element in the SVG specification. It's the usual Perlin noise.
+/// See: https://www.w3.org/TR/SVG11/filters.html#feTurbulenceElement
+/// The `octave_offsets` parameter of `turbulence` was added after porting.
```
Per the [W3C Document License](https://www.w3.org/Consortium/Legal/2015/doc-license), let's add this notice:
`Copyright © 2015 W3C® (MIT, ERCIM, Keio, Beihang). This software or document includes material copied from or derived from https://www.w3.org/TR/SVG11/filters.html#feTurbulenceElement.`

Herschel reviewed (changes requested):
Thanks! Feels a little odd to have `turbulence` in the top-level module. Let's either move it into `bitmap_data.rs` or make a new `bitmap` module on the top-level and move it there. (The latter might make a good choice because most of the bitmap algos will be shared by AVM2, as well.)

torokati44 commented:
Thanks for the review! Added the copyright notice, and moved `turbulence.rs` to a new toplevel module.
(I hope you meant it this way. To be honest, initially I just put that file _somewhere_ and hammered it until it worked, and hoped that you would direct as to where it's place should be - and so it was... :) )

Herschel commented:
Thanks! I think eventually we'll move most of the `avm1::object::bitmap_data` stuff into this module as well, since it will all be shared by AVM2.
